libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -0,0 +1,572 @@
+      ast::identifier{strip_prefix(rw_plugin->name()), location::unknown},
+    }};
+    if constexpr (not is_loading) {
+      pipeline_argument->inner.body.emplace_back(
+        ast::invocation{std::move(io_ent), {}});
+    }
+    if (compression_plugin) {
+      auto compression_ent = ast::entity{std::vector{
+        ast::identifier{strip_prefix(compression_plugin->name()),
+                        location::unknown},
+      }};
+      pipeline_argument->inner.body.emplace_back(ast::invocation{
+        std::move(compression_ent),
+        {ast::constant{std::string{compression_name}, location::unknown}}});
+    }
+    if constexpr (is_loading) {
+      pipeline_argument->inner.body.emplace_back(
+        ast::invocation{std::move(io_ent), {}});
+    }
+    TENZIR_ASSERT(resolve_entities(pipeline_argument->inner, ctx));
+  }
+  TENZIR_TRACE("{} operator: final pipeline        :", traits::operator_name);
+  for (auto& arg : inv.args) {
+    TENZIR_TRACE("    {:?}", arg);
+  }
+  if (io_properties.accepts_pipeline) {
+    return io_plugin->make(std::move(inv), std::move(ctx));
+  } else {
+    auto compiled_pipeline = pipeline{};
+    if (pipeline_argument) {
+      TRY(compiled_pipeline, compile(std::move(pipeline_argument->inner), ctx));
+      TENZIR_TRACE("{} operator: compiled pipeline ops : {}",
+                   traits::operator_name, compiled_pipeline.operators().size());
+      inv.args.pop_back();
+    }
+    TRY(auto io_op, io_plugin->make(std::move(inv), ctx));
+    if constexpr (is_loading) {
+      compiled_pipeline.prepend(std::move(io_op));
+    } else {
+      compiled_pipeline.append(std::move(io_op));
+    }
+    return std::make_unique<pipeline>(std::move(compiled_pipeline));
+  }
+}
+
+class from_plugin2 final : public virtual operator_factory_plugin {
+public:
+  constexpr static auto docs = "https://docs.tenzir.com/tql2/operators/from";
+  auto name() const -> std::string override {
+    return "tql2.from";
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    if (inv.args.empty()) {
+      diagnostic::error("expected positional argument `<path/url/events>`")
+        .primary(inv.self)
+        .docs(docs)
+        .emit(ctx);
+      return failure::promise();
+    }
+    auto& expr = inv.args[0];
+    TRY(auto value, const_eval(expr, ctx));
+    auto events = std::vector<record>{};
+    using ret = std::variant<bool, failure_or<operator_ptr>>;
+    const auto extract_single_event = [&](record& event) -> ret {
+      events.push_back(std::move(event));
+      return true;
+    };
+    auto result = match(
+      value, extract_single_event,
+      [&](std::string& path) -> ret {
+        return create_pipeline_from_uri<true>(path, std::move(inv), ctx, docs);
+      },
+      [&]<typename T>(T& value) -> ret
+      // requires(not std::same_as<T, record>)
+      {
+        auto t = type::infer(value);
+        diagnostic::error("expected a URI, or a record")
```

Comment:
Maybe "string" instead of or in addition to "URI", as the underlying reason is a type mismatch?